### PR TITLE
RFC: Support for interpolated strings in D

### PR DESCRIPTION
**Pull request by jcd**
Support for interpolated strings like e.g. C# has. It will let you write code like:

```
void main()
{
  auto name = "John Doe";
  auto halfAge = 42;
  writeln( $"The age of {name} is {2 * halfAge}");
}
```

This PR is a request for comments on the approach taken and to see if this is something that has a chance of being included.

It needs the addition of the following function to object.d :

```
string _interp_text(T...)(T args)
{
    static if(T.length == 0)
        return null;

    // Check if phobos std.conv is present and use if so
    // else make a best effort to convert to string.
    static if (__traits(compiles, { import std.conv; return 0; }()))
    {
        static import std.conv;
        return std.conv.text(args);
    }
    else static if (__traits(compiles, args[0].toString()))
        return args[0].toString() ~ _interp_text(args[1..$]);
    else static if (is(args[0] == string))
        return args[0] ~ _interp_text(args[1..$]);
    else
        return "?" ~ _interp_text(args[1..$]);
}
```

### CODE REVIEW

**Comment by WalterBright** [comment]
newsgroup discussion:
http://www.digitalmars.com/d/archives/digitalmars/D/Interpolated_strings_299875.html

File: src/ddmd/parse.d [hunk 1]
```diff
@@ -7051,12 +7147,43 @@ final class Parser : Lexer
...
+                    error(result.peekString());
+                    mem.xfree(result.extractData());
+                    goto Lerr;
+                }
+                else
+                {
+                    stringExpStr = result.peekSlice();
+                }
+
+                auto s = cast(char*)mem.xmalloc(stringExpStr.length);
```
Comment (UplinkCoder) [inline src/ddmd/parse.d]: use allocmemory here, not the nasty xmalloc which costs atlot.
and is only for temporary strings which are supposed to be freed.
please just alloc the buffer once and reuse it if possible

**Comment by WalterBright** [comment]
Overview of existing methods:
https://en.wikipedia.org/wiki/String_interpolation

**Comment by UplinkCoder** [comment]
My first impression is that the syntax is nasty.
using the dollar operator is quite dicey.

**Comment by UplinkCoder** [comment]
I'd rather see ` #"his name was ${name}"`  ...

**Comment by ibuclaw** [comment]
> My first impression is that the syntax is nasty.
using the dollar operator is quite dicey.

Indeed, I would much prefer `%(var)` or at worst `%{var}` just to keep in line with formatting rules of std format.

**Comment by wilzbach** [comment]
As the NG discussion led to no consensus, it would be great if we could work on comparing possible syntax variants as part of a DIP, s.t. a sophisticated decision can be made.

It can be inspired by 

### Accepted proposals in other languages

- [PEP 498 - Literal String Interpolation in Python](https://www.python.org/dev/peps/pep-0498)
- [Template Literal Proposal for ES6](https://tc39.github.io/proposal-template-literal-revision/)
- [String interpolation](https://en.wikipedia.org/wiki/String_interpolation)

### Language documentations

- [RosettaCode](https://www.rosettacode.org/wiki/String_interpolation_(included))
- [C#](https://docs.microsoft.com/en-us/dotnet/articles/csharp/language-reference/keywords/interpolated-strings)
- [CoffeeScript](https://coffeescript-cookbook.github.io/chapters/strings/interpolation)
- [Dart](https://www.dartlang.org/guides/language/language-tour#strings)
- [Groovy](http://docs.groovy-lang.org/latest/html/documentation/#_string_interpolation)
- [Haxe](https://haxe.org/manual/lf-string-interpolation.html)
- [JS](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Template_literals)
- [Perl](http://perl101.org/strings.html)
- [PHP](http://php.net/manual/en/language.types.string.php)
- [Ruby](http://ruby-for-beginners.rubymonstas.org/bonus/string_interpolation.html)
- [Scala](http://docs.scala-lang.org/overviews/core/string-interpolation.html)
- [Swift](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/StringsAndCharacters.html#//apple_ref/doc/uid/TP40014097-CH7-ID292)

**Comment by andralex** [comment]
So what's the status of this? Do we have a champion for a DIP?

**Comment by wilzbach** [comment]
As #7988 seems more active, I'm closing this in favor of #7988

**Pull request closed without merging**